### PR TITLE
Fix crash in Daemon::answer_client_requests

### PR DESCRIPTION
**Pull request by tombailo**
iceccd was regularly crashing on my PC, and investigation showed the
cause to be using an invalid iterator in Daemon::answer_client_requests.
This function calls create_env_finished for each entry in
native_environments, but create_env_finish will sometimes erase the
entry, leaving the calling function with an invalid iterator.

This commit fixes the crash by moving the erase() up to
answer_client_requests where it can be done safely

### CODE REVIEW

**Comment by HenryMiller1** [comment]
I'd like to see some tests for this situation if you can figure out how to write them

File: daemon/main.cpp [hunk 1]
```diff
@@ -1884,10 +1884,15 @@ int Daemon::answer_client_requests()
...
+                 it != native_environments.end(); ) {
                 if (it->second.create_env_pipe && FD_ISSET(it->second.create_env_pipe, &listen_set)) {
-                    create_env_finished(it->first);
+                    if(!create_env_finished(it->first))
+                    {
+                        native_environments.erase(it++);
+                        continue;
+                    }
                 }
+                ++it;
```
Comment (HenryMiller1) [inline daemon/main.cpp]: Why did you move this out of the for loop?

Comment (HenryMiller1) [inline daemon/main.cpp]: That is badly worded - why did you move this out of the for command to a line of its own?

Comment (johnmiked15) [inline daemon/main.cpp]: Because `erase()` invalidates the iterator:

```c++
   native_environments.erase(it++);
```
`erase()` is called with the value of the current `it`, but is incremented before the next loop iteration since it uses post-increment. If it's not getting erased, then it needs to be incremented anyway, hence moving it out the for loop. You can't call `erase(it)` within the body of the for loop, because then the `it++` operation that happens after each loop iteration will be using an invalidated `it`.

Comment (HenryMiller1) [inline daemon/main.cpp]: I see, I missed that.  (and had to double check the rules on iterator invalidation for std::map - looks like this is acceptable.

Doesn't this skip the next element in the map if an erase happens?

Comment (johnmiked15) [inline daemon/main.cpp]: No...this is a pretty standard loop that erases elements based on a condition. It looks fine to me.

Comment (HenryMiller1) [inline daemon/main.cpp]: Ahh, the next line is continue...  My mistake.

**Comment by HenryMiller1** [comment]
The change looks good, but I'd like to see some test for this situation so it isn't reintroduced.

**Comment by tombailo** [comment]
Hi all, sorry for the radio silence on this. I'm happy to add some tests for this, but can you point me towards some suitable examples? I can't see any tests for the daemon in the tests/ directory

**Comment by HenryMiller1** [comment]
I'm not sure if the current test approach lets you access details like this.  I just want contributors to think about testing.
